docs/okeps/okep-5259-no-overlay.md

```diff
@@ -0,0 +1,1054 @@
+
+Changes to BGP behavior are necessary to import node subnet routes into the
+gateway router and the host routing table of each node. A new `encapsulation`
+field will be introduced to the NAD spec that indicates the encapsulation of the
+network. If a network is operating in no-overlay mode, this field will be set to
+`none`. We don't support users modifying this field in a NAD manually to enable
+no-overlay mode. This information can then be passed into the NetInfo object and
+utilized by both the RouteAdvertisement controller and the route import
+controller.
+
+An example OVN-Kubernetes NAD may look like:
+
+```yaml
+apiVersion: k8s.cni.cncf.io/v1
+kind: NetworkAttachmentDefinition
+metadata:
+  name: l3-network
+  namespace: default
+spec:
+  config: |
+    {
+            "cniVersion": "0.3.1",
+            "name": "l3-network",
+            "type": "ovn-k8s-cni-overlay",
+            "topology":"layer3",
+            "mtu": 1500,
+            "netAttachDefName": "default/l3-network",
+            "role": "primary",
+            "encapsulation": "none"
+    }
+```
+
+##### Host Routing Tables
+
+To [maintain isolation between
+UDNs](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5140), node BGP
+speakers do not accept routes to node subnets by default. However, to route
+east-west traffic in no-overlay mode, these routes must be imported to the host
+routing table. We need to update the RouteAdvertisement controller which is
+responsible for generating the per-node FRRConfiguration instance. For a network
+operating in no-overlay mode, the RouteAdvertisement controller shall generate
+the FRRConfiguration that accepts the routes to pod node subnets explicitly.
+
+##### Gateway Router
+
+Currently, BGP routes to other node subnets are not imported into the gateway
+router, which ensures node-to-pod traffic traverses the OVN logical topology.
+However, we need the routes to be imported into the gateway router in no-overlay
+mode. We need to update the route import controller to check whether a network
+is in no-overlay mode and import the routes from the host routing table if so.
+
+#### Conditional SNAT for Egress Traffic
+
+The OVN-Kubernetes BGP feature currently disables SNAT for all pod egress
+traffic if the pod network is advertised. The only exception is for traffic
+destined to other nodes in the cluster (pod-to-other-node), to ensure nodePort
+services can be accessed across networks. This egress traffic will remain SNATed
+to the node IP before leaving the node.
+
+However, this approach assumes that all external destinations can route back to
+the pod IPs, which is often not true for destinations reached via non-BGP routes
+(like a default internet gateway). So, instead of disabling SNAT for all pod
+egress traffic, we shall do the following:
+
+* For egress traffic forwarded via a BGP route, SNAT shall be disabled.
+* For egress traffic forwarded via a non-BGP route, the egress traffic shall be
+  SNATed to the node IP.
```

Comment:
This seams really complicated to pull off. Is it really a hard requirement? Could we just expose a field in the RA that says "SNAT egress traffic" or something? If true, then we only add pod subnets to the address set and skip SNAT if they are in there (positive match), all other traffic is SNAT'ed.